wsdl_docs.inc

```diff
@@ -606,32 +617,63 @@ function _wsdl_docs_service_import($name) {
   else {
     return "No service found with this name.";
   }
-  // Import via file.
-  // Manipulate $_FILES to match form version.
-  $_FILES['files'] = $_FILES['soap_api_definition'];
-  // Confirm file is of correct type.
-  $tmp = explode(".", $_FILES['soap_api_definition']['name']);
-  $ext = array_pop($tmp);
-  if ($ext != 'wsdl') {
-    return services_error(t('incorrect file format'), 500);
+  // Create $_FILES['files'] from array of file information provided by REST request.
+  $_FILES['files'] = array();
+  foreach ($_FILES['soap_api_definition'] as $file_key => $values) {
+    foreach ($values as $key => $value) {
+      $_FILES['files'][$key][$file_key] = $value;
+    }
+  }
+
+  // First, confirm file(s) of correct extension.
+  foreach ($_FILES['files'] as $key => $file) {
+    $tmp = explode(".", $file['name']);
+    $ext = array_pop($tmp);
+    if (!in_array($ext, array('xsd', 'wsdl'))) {
+      return t('incorrect file format');
+    }
   }
-  // Save wsdl file.
-  $wsdl = file_get_contents($_FILES['soap_api_definition']['tmp_name']);
+
+  // Second, save file to selected directory.
+  // @todo - do we want to save each set of files in a subdirectory that is the
+  // service name and id to make sure we don't get overwriting of files with the
+  // same name?
   $directory = WSDL_DIRECTORY;
   file_prepare_directory($directory, FILE_CREATE_DIRECTORY);
-  $filename = WSDL_DIRECTORY . $name . '-' . $service_id . '.wsdl';
-  file_unmanaged_save_data($wsdl, $filename, FILE_EXISTS_REPLACE);
-  // Save new WSDL file to existing WSDL service entity.
-  $service->url = $filename;
-  $service->save();
-  // Load file contents.
-  $DOMDocument = new DOMDocument();
-  $loaded = $DOMDocument->loadXML($wsdl);
-  if (!$loaded) {
-    return 'ERROR input file is not XML';
+  foreach ($_FILES['files'] as $key => $file) {
+    $wsdl = file_get_contents($file['tmp_name']);
+    // For the main WSDL/XSD file that refers to other files, name it
+    // [service name]-[service id].wsdl or .xsd, depending on file's extension.
+    if ($key == 0) {
+      $tmp = explode(".", $file['name']);
+      $ext = array_pop($tmp);
+      $tmp_filename = WSDL_DIRECTORY . $service->name . '-' . $service->id . '.' . $ext;
+      $main_filename = $tmp_filename;
+    }
+    // For any additional files, keep the name the same so references from
+    // the main file still work.
+    else {
+      $tmp_filename = WSDL_DIRECTORY . $file['name'];
+    }
+    file_unmanaged_save_data($wsdl, $tmp_filename, FILE_EXISTS_REPLACE);
```

Comment:
This also seems dangerous.  We're taking an externally specified path with minimal validation, retrieving its contents and then writing that unvalidated information to a live file with a predictable filename/location.

This is particularly an issue if the site using this code is running an older version of PHP that were more vulnerable to object injection attacks.  In this case, say the user is able to upload a "WSDL" file that contains specially formatted object data.  When that data is retrieved file `file_get_contents` or `DOMDocument`, PHP would make an attempt to deserialize that data into an object which allowed the hacker to inject malicious code into the current execution thread.

Reply:
For this, before any file is saved it is checked to see that the XML is valid (line 655). Once the file is saved, it is then checked to see if the WSDL is valid. If not the file is deleted (line 680).